### PR TITLE
Portable bundle manager in Samza Runner

### PR DESCRIPTION
**Pull request by ryucc**
Implement PortableBundleManager in Samza Runner

Changes from ClassicBundleManager
1. The state variables need to be updated atomically. Now they are bundled together using a synchronized block instead of individual atomics.
2. Change the atomic variables to primitives, since they are already in a synchronized block.
3. Remove the future collector and future watermarks. Now we can send watermarks once the bundle is finished.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/PortableBundleManager.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  // Number of bundles that are in progress but not yet finished
+  private transient long pendingBundleCount;
+  // Denotes the start time of the current active bundle
+  private transient long bundleStartTime;
+  // Denotes if there is an active in progress bundle. Note at a given time, we can have multiple
+  // bundle in progress.
+  // This flag denotes if there is a bundle that is current and hasn't been closed.
+  private transient boolean isBundleStarted;
+  // Holder for watermark which gets propagated when the bundle is finished.
+  private transient Instant bundleWatermarkHold;
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/PortableBundleManager.java:65]: Is the intention behind using transient to communicate these fields are not meant for serialization if at ll this class is serialized in the future? 
If not, can you comment what is the rationale behind this style?

File: runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  public void testProcessWatermarkWhenBundleNotStarted() {
+    Instant watermark = new Instant();
+    portableBundleManager =
+        new PortableBundleManager<>(bundleProgressListener, 4, 1, bundleTimerScheduler, TIMER_ID);
+    portableBundleManager.processWatermark(watermark, emitter);
+    verify(bundleProgressListener, times(1)).onWatermark(eq(watermark), eq(emitter));
+  }
+
+  @Test
+  public void testQueueWatermarkWhenBundleNotStarted() {
```
Comment (mynameborat) [inline runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java:134]: s/testQueueWatermarkWhenBundleNotStarted/testQueueWatermarkWhenBundleStarted?

File: runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    portableBundleManager.tryStartBundle();
+    KeyedTimerData<Void> keyedTimerData = mock(KeyedTimerData.class);
+    TimerInternals.TimerData timerData = mock(TimerInternals.TimerData.class);
+    when(keyedTimerData.getTimerData()).thenReturn(timerData);
+    when(timerData.getTimerId()).thenReturn(TIMER_ID);
+
+    portableBundleManager.processTimer(keyedTimerData, emitter);
+    verify(bundleProgressListener, times(1)).onBundleFinished(any());
+  }
```
Comment (mynameborat) [inline runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java:160]: can we validate if we are scheduling the next timer as well?

**Review by mynameborat — COMMENTED** [review]
Minor comments/questions. Looks good otherwise.

File: runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java (hunk 2 above)
Comment (ryucc) [reply runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java:134]: Fixed.

File: runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java (hunk 3 above)
Comment (ryucc) [reply runners/samza/src/test/java/org/apache/beam/runners/samza/runtime/PortableBundleManagerTest.java:160]: Added

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/PortableBundleManager.java (hunk 1 above)
Comment (ryucc) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/PortableBundleManager.java:65]: I'll remove it. Doesn't look like this class can be serialized. (or not needed at the moment.) I was copying the classic version, but the Scheduler and Listener are probably not serializable as well.

**Comment by tvalentyn** [comment]
@ryucc do you have a committer in mind familiar with Samza runner who you are working with?

**Comment by ryucc** [comment]
@tvalentyn I'm working with @xinyuiscool and @mynameborat
Thanks for checking in!

**Comment by tvalentyn** [comment]
np. consider adding a group of experts for `samza` label in https://github.com/apache/beam/blob/3e3c7a4e0390dfc380f8f65f137ecdb3e88a2f9d/.github/REVIEWERS.yml

**Comment by kennknowles** [comment]
I think we need some registered reviewers for Samza!

**Comment by kennknowles** [comment]
@xinyuiscool would you be able to sign up and find friends to be in the OWNERS file so that our review bot can ping when there is a change to the samza runner?

**Comment by xinyuiscool** [comment]
> @xinyuiscool would you be able to sign up and find friends to be in the OWNERS file so that our review bot can ping when there is a change to the samza runner?

Will update the OWNERS file very soon. Sorry about the delay.

**Review by xinyuiscool — APPROVED** [review]
LGTM!

**Pull request merged**